File: OJ/Leetcode/Algorithm/767. Reorganize String.java
class Solution {
public String reorganizeString(String S) {
        int n = S.length();
        if(n <= 1) 
            return S;
        int[] str = new int[26];
        int hf = 0;
        for(char a: S.toCharArray())
            if(hf < ++str[a-'a']) 
                hf = str[a-'a'];
        if(hf*2-1>n) return "";
        char[] res=new char[n];
        int odd=0, even=1;
        for(int i=0;i<26;i++)
        {
            while(str[i]>0 && str[i]<n/2+1 && even<n)
            {
                res[even]= (char)(i+'a');
                even+=2;
                str[i]--;
            }
            while(str[i]>0)
            {
                res[odd]=(char)(i+'a');
                odd+=2;
                str[i]--;
            }
        }
        return new String(res);
    }
}

File: OJ/Leetcode/Algorithm/767. Reorganize String.cpp
class Solution {
public:
    string reorganizeString(string S) {
        int maxx = 0, c;
        int n = S.length();
        map<int, int> mp;
        for(int i = 0 ; i < n; i++)
        {
            mp[S[i]]++;
            if(mp[S[i]] > maxx)
            {
                maxx = mp[S[i]];
                c = S[i];
            }
        }
        if(maxx > (n+1)/2)
            return "";
        int even = 0, odd = 1;
        string ans(S.length(), ' ');
        for(int i = 0; i < maxx; i++, even+=2)
            ans[even] = c;
        mp[c] = 0;

        for(int i = 'a'; i <= 'z'; i++)
        {
            while(mp[i] > 0)
            {
                if(even < n)
                {
                    ans[even] = i;
                    even+=2;
                }
                else
                {
                    ans[odd] = i;
                    odd+=2;
                }
                mp[i]--;
            }
        }
        return ans;
    }
};
